core/chains/terra/monitor/balance.go

```diff
@@ -0,0 +1,110 @@
+package monitor
+
+import (
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/smartcontractkit/chainlink-terra/pkg/terra"
+	"github.com/smartcontractkit/chainlink-terra/pkg/terra/client"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services"
+	"github.com/smartcontractkit/chainlink/core/services/keystore"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+func NewBalanceMonitor(chainID string, cfg terra.Config, lggr logger.Logger, ks keystore.Terra) services.Service {
+	return &balanceMonitor{
+		chainID: chainID,
+		cfg:     cfg,
+		lggr:    lggr.Named("BalanceMonitor"),
+		ks:      ks,
+		stop:    make(chan struct{}),
+		done:    make(chan struct{}),
+	}
+}
+
+type balanceMonitor struct {
+	utils.StartStopOnce
+	chainID string
+	cfg     terra.Config
+	lggr    logger.Logger
+	ks      keystore.Terra
+
+	stop, done chan struct{}
+}
+
+func (b *balanceMonitor) Start() error {
+	return b.StartOnce("TerraBalanceMonitor", func() error {
+		go b.monitor()
+		return nil
+	})
+}
+
+func (b *balanceMonitor) Close() error {
+	return b.StopOnce("TerraBalanceMonitor", func() error {
+		close(b.stop)
+		<-b.done
+		return nil
+	})
+}
+
+var promTerraBalance = promauto.NewGaugeVec(
+	prometheus.GaugeOpts{Name: "terra_balance", Help: "Terra account balances"},
+	[]string{"account", "terraChainID", "denomination"},
+)
+
+func (b *balanceMonitor) updateProm(acc sdk.AccAddress, bal *sdk.DecCoin) {
+	balF, err := bal.Amount.Float64()
+	if err != nil {
+		b.lggr.Errorw("Failed to convert balance to float", "err", err)
+	}
+	promTerraBalance.WithLabelValues(acc.String(), b.chainID, bal.GetDenom()).Set(balF)
+}
+
+func (b *balanceMonitor) monitor() {
+	defer close(b.done)
+
+	tick := time.After(utils.WithJitter(b.cfg.BlockRate()))
+
+	for {
+		select {
+		case <-b.stop:
+			return
+		case <-tick:
+			b.updateBalances()
+			tick = time.After(utils.WithJitter(b.cfg.BlockRate()))
+		}
+	}
+}
+
+func (b *balanceMonitor) updateBalances() {
+	keys, err := b.ks.GetAll()
+	if err != nil {
+		b.lggr.Errorw("Failed to get keys", "err", err)
+		return
+	}
+	var rw client.ReaderWriter
+	for _, k := range keys {
+		select {
+		case <-b.stop:
+			return
+		default:
+		}
+		acc := sdk.AccAddress(k.PublicKey().Address())
+		bal, err := rw.Balance(acc, "uluna") //TODO async to prevent blocking shutdown?
```

Comment:
This should be receiving a context, internally it is using `context.Background()` 😓

Reply:
IIRC somewhere even further down doesn't respect the context, and we've had to work around it already. I'll revisit this though to be sure.